CodeGeneration/Sources/SyntaxSupport/Child.swift

```diff
@@ -244,32 +263,118 @@ public class Child: NodeChoiceConvertible {
     documentation: String? = nil,
-    isOptional: Bool = false
+    isOptional: Bool = false,
+    newerChildPath: [Child] = []
   ) {
     precondition(name.first?.isLowercase ?? true, "The first letter of a child’s name should be lowercase")
-    precondition(
-      deprecatedName?.first?.isLowercase ?? true,
-      "The first letter of a child’s deprecatedName should be lowercase"
-    )
     self.name = name
-    self.deprecatedName = deprecatedName
+    self.newerChildPath = newerChildPath
     self.kind = kind
     self.experimentalFeature = experimentalFeature
     self.nameForDiagnostics = nameForDiagnostics
     self.documentationSummary = SwiftSyntax.Trivia.docCommentTrivia(from: documentation)
     self.documentationAbstract = String(documentation?.split(whereSeparator: \.isNewline).first ?? "")
     self.isOptional = isOptional
   }
+
+  /// Create a node that is a copy of the last node in `newerChildPath`, but
+  /// with modifications.
+  init(renamingTo replacementName: String? = nil, newerChildPath: [Child]) {
+    let other = newerChildPath.last!
+
+    self.name = replacementName ?? other.name
+    self.newerChildPath = newerChildPath
+    self.kind = other.kind
+    self.experimentalFeature = other.experimentalFeature
+    self.nameForDiagnostics = other.nameForDiagnostics
+    self.documentationSummary = other.documentationSummary
+    self.documentationAbstract = other.documentationAbstract
+    self.isOptional = other.isOptional
+  }
+
+  /// Create a child for the unexpected nodes between two children (either or
+  /// both of which may be `nil`).
+  convenience init(forUnexpectedBetween earlier: Child?, and later: Child?, newerChildPath: [Child] = []) {
+    let name = switch (earlier, later) {
+    case (nil, let later?):
+      "unexpectedBefore\(later.name.withFirstCharacterUppercased)"
+    case (let earlier?, nil):
+      "unexpectedAfter\(earlier.name.withFirstCharacterUppercased)"
+    case (let earlier?, let later?):
+      "unexpectedBetween\(earlier.name.withFirstCharacterUppercased)And\(later.name.withFirstCharacterUppercased)"
+    case (nil, nil):
+      "unexpected"
+    }
+
+    self.init(
+      name: name,
+      kind: .collection(kind: .unexpectedNodes, collectionElementName: name.withFirstCharacterUppercased),
+      experimentalFeature: earlier?.experimentalFeature ?? later?.experimentalFeature,
+      nameForDiagnostics: nil,
+      documentation: nil,
+      isOptional: true,
+      newerChildPath: newerChildPath
+    )
+  }
+}
+
+extension Child: Hashable {
+  public static func == (lhs: Child, rhs: Child) -> Bool {
+    lhs === rhs
+  }
+  
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(ObjectIdentifier(self))
+  }
 }
+
+extension Child {
+  /// A set of changes to the children that were all made simultaneously. The
+  /// key is the *new* name of the child; any names in the value are old names.
+  public typealias ChangeSet = [String: Refactoring]
```

Comment:
I think I like the idea of the change set here but it’s a little hard to review because it’s combined with the `@abi` attribute changes. Do you think you could pull out the refactoring to `childHistory` into a separate, preliminary PR?

Reply:
Done here: #2888